### ISSUE TITLE
agent failed, code: 23

### ISSUE THREAD

**Issue by kohldampfer** [fossology/fossology#1076]
During a check with Fossology, I get this error message:

> 2018-04-24 08:37:56 wget_agent [0] :: JOB[752].wget_agent[34874.localhost]: "FATAL wget_agent.c.521: [wget_agent.c:521] Error exit: 1, CMD:svn export https://XXXX  --username XXXX --password XXXX /srv/fossology/repository/c62111288937/wget/wget.34874.dir --no-auth-cache >/dev/null 2>&1"
>
> 2018-04-24 08:37:56 wget_agent [0] :: JOB[752].wget_agent[34874.localhost]: "FATAL wget_agent.c.525: please make sure the URL of repo is correct, also add correct proxy for your version control system, command is:svn export https://XXXX  --username XXXX --password XXXX /srv/fossology/repository/c62111288937/wget/wget.34874.dir --no-auth-cache >/dev/null 2>&1, GlobalTempFile is:/srv/fossology/repository/c62111288937/wget/wget.34874, rc is:256. "
>
> 2018-04-24 08:37:56 wget_agent [0] :: JOB[752].wget_agent[34874.localhost]: "FATAL main.c.243: upload 85 File retrieval failed: uploadpk=85 tempfile=/srv/fossology/repository/c62111288937/wget/wget.34874 URL=https://XXXX Type=SVN"
>
> 2018-04-24 08:37:56 wget_agent [0] :: JOB[752].wget_agent[34874.localhost]: agent failed with error code 23
>
> 2018-04-24 08:37:56 wget_agent [0] :: JOB[752].wget_agent[34874.localhost]: agent failed, code: 23

Here, I use the functionality of **From Version Control System**. I do this in an automatic manner, i.e. with a script, which sends a HTTP request. This script basically works, because there are several projects which call the same script. I also checked, if URL / username / password is correct.

Is there a way to  check, what's the problem here is? Or is here a way to get more detailled output?

**Comment by GMishx** [comment]
Are you using FOSSology behind proxy? If yes, try changing the proxy settings in `fossology.conf` and restarting the scheduler.
If the problem persists, can you please tell the type of VCS you are using (git, svn, mercurial, etc.)?

**Comment by kohldampfer** [comment]
> Are you using FOSSology behind proxy?

No. I use Fossology withing a docker conatiner, but this shouldn't be a problem. There are other repositories, which can be scanned without any problems and the SVN server is the same. I checked the address and it is correct.

> If the problem persists, can you please tell the type of VCS you are using (git, svn, mercurial, etc.)?

svn

**Comment by mcjaeger** [comment]
(from discussion) actually it seems like svn is installed, URL can be reached, but the call times out. @kohldampfer error is still persisting?

**Comment by kohldampfer** [comment]
@mcjaeger I have to check. I will respond later.

**Comment by kohldampfer** [comment]
@mcjaeger Sorry for the late answer. I just got the same error message again when I tried to use Fossology again. My assumption is, that it has something to do with the SSL certificate of the subversion server. The certificate is a self-signed one and wget / svn are not accepting these sort of certificates by default.

**Comment by kohldampfer** [comment]
I want to write down, what I just did and what results I had. First, I changed the file wget_agent.c (https://github.com/fossology/fossology/blob/master/src/wget_agent/agent/wget_agent.c) or rather the line 942 from this

```
      sprintf(command, "svn export %s %s %s --no-auth-cache >/dev/null 2>&1", GlobalURL, GlobalParam, TempFileDirectory);
```
to this
```
      sprintf(command, "svn export %s %s %s --no-auth-cache 2>&1", GlobalURL, GlobalParam, TempFileDirectory);
```

So, with this I can see the real errors. Afterwards, I removed all old docker container and images and recreated them with `docker-compose up -d`.

Then I tried to let Fossology download a repo and it failed (foreseeable) because of the self-signed certificate:

> 2018-10-05 14:30:15 wget_agent [0] :: JOB[38].wget_agent[1662.localhost]: "svn: E230001: Unable to connect to a repository at URL 'https://svn.some.host/repo'"
> 2018-10-05 14:30:15 wget_agent [0] :: JOB[38].wget_agent[1662.localhost]: "svn: E230001: Server SSL certificate verification failed: issuer is not trusted"

So I tried to add the self-signed certificate to the container. I did the following steps for the container `fossology_web_1` and `fossology_scheduler_1`:

1. Downloaded the certificate and update certificates on the container
```
$ openssl s_client -connect svn.some.host:443 > host.crt
# adjusting host.crt, so there is only the part from `-BEGIN CERTIFICATE-` till `-END CERTIFICATE-`
# details are here: https://serverfault.com/a/192731
$ cp -v host.crt /usr/local/share/ca-certificates
$ update-ca-certificates
```

2. Let svn accept the certificate, i.e. I called `svn info https://svn.some.host/repo`:
```
$ svn info https://svn.some.host/repo
Error validating server certificate for 'https://svn.some.host:443':
 - The certificate is not issued by a trusted authority. Use the
   fingerprint to validate the certificate manually!
Certificate information:
 - Hostname: svn.some.host
 - Valid: from Mar 20 11:09:30 2018 GMT until Mar 19 11:09:30 2021 GMT
 - Issuer:
 - Fingerprint: 
(R)eject, accept (t)emporarily or accept (p)ermanently? p
```

Here, I removed the part after `Fingerprint`. I answered the question with `p`. I did this as user `root` and as user `fossy` (`su - fossy`).

But I still get this error message in Fossology, when I want to start a new job:

> 2018-10-05 14:30:15 wget_agent [0] :: JOB[38].wget_agent[1662.localhost]: "svn: E230001: Server SSL certificate verification failed: issuer is not trusted"

**Comment by szrusko** [comment]
any solution for this problem?

**Comment by mcjaeger** [comment]
In case it is a similar cause as with git - the used certificate is not trusted by your system - this stack overflow article could help:

https://stackoverflow.com/questions/31544633/how-to-trust-an-expired-ssl-certificate-with-subversion
